lib/python-indent.coffee

```diff
@@ -42,31 +47,29 @@ module.exports = PythonIndent =
 
     # Get base variables
     row = editor.getCursorBufferPosition().row
-    previousLine = editor.buffer.lineForRow(row - 1)
-    return unless row
-
-    # Based on configuration, branch to corresponding logic
-    if (matchOpeningDelimiter = PythonIndent.openingDelimiterIndentRegex.exec previousLine) isnt null
-      PythonIndent.indentOnOpeningDelimiter(editor, row, previousLine, matchOpeningDelimiter)
-    else if PythonIndent.hangingIndentRegex.test previousLine
-      PythonIndent.indentHanging(editor, row, previousLine)
-    else if PythonIndent.openingDelimiterUnindentRegex.test previousLine
-      PythonIndent.unindentOnOpeningDelimiter(editor, row, previousLine)
-
-  indentOnOpeningDelimiter: (editor, row, previousLine, matchOpeningDelimiter) ->
-    # Get index of opening delimiter.
-    # The indent should be one whitespace character past this (sorry, no tabs)
-    indentColumn = previousLine.lastIndexOf(matchOpeningDelimiter[1]) + 1
+    col = editor.getCursorBufferPosition().column
+
+    # Parse the entire file up to the current point, keeping track of brackets
+    lines = editor.getTextInBufferRange([[0,0], [row, col]]).split('\n')
+
+    # A stack of [row, col] pairs describing where open brackets are
+    bracketStack = PythonIndent.parseLines(lines)
+
+    return unless bracketStack.length
+
+    # bracketStack.pop()[1] is the column where the bracket was, so need to bump by one
+    indentColumn = bracketStack.pop()[1] + 1
 
     # Get tab length for context
     tabLength = editor.getTabLength()
 
     # Calculate soft-tabs from spaces (can have remainder)
-    tabs = (indentColumn / tabLength) - editor.indentationForBufferRow row
-    rem = indentColumn % tabLength
+    tabs = (indentColumn / tabLength)
+    rem = ((tabs - Math.floor(tabs)) * tabLength)
 
     # If there's a remainder, `editor.buildIndentString` requires the tab to
     # be set past the desired indentation level, thus the ceiling.
+    # TODO: is the conditional necessary? Maybe to avoid floating point errors?
```

Comment:
Yes, this is because of the possibility that `rem` could be 0, which is totally acceptable. I remember this when testing through everything, though I can't remember the exact scenario.